source/validate.h

```diff
@@ -84,91 +103,284 @@ enum class FunctionDecl {
 };
 
 class ValidationState_t;
+class Function;
+
+// This class represents a basic block in a SPIR-V module
+class BasicBlock {
+ public:
+  /// Constructor for a BasicBlock
+  ///
+  /// @param[in] id The ID of the basic block
+  /// @param[in] module A reference of the module of the basic block
+  BasicBlock(uint32_t id);
+
+  /// Returns the id of the BasicBlock
+  uint32_t get_id() const { return id_; }
+
+  /// Returns the predecessors of the BasicBlock
+  const std::vector<BasicBlock*>& get_predecessors() const {
+    return predecessors_;
+  }
+
+  /// Returns the predecessors of the BasicBlock
+  std::vector<BasicBlock*>& get_predecessors() { return predecessors_; }
+
+  /// Returns the successors of the BasicBlock
+  const std::vector<BasicBlock*>& get_successors() const { return successors_; }
+
+  /// Returns the successors of the BasicBlock
+  std::vector<BasicBlock*>& get_successors() { return successors_; }
+
+  /// Sets the immedate dominator of this basic block
+  ///
+  /// @param[in] dom_block The dominator block
+  void SetImmediateDominator(BasicBlock* dom_block);
+
+  /// Returns the immedate dominator of this basic block
+  BasicBlock* GetImmediateDominator();
+
+  /// Returns the immedate dominator of this basic block
+  const BasicBlock* GetImmediateDominator() const;
+
+  /// Sets the @p next BasicBlock as the successor of this BasicBlock
+  void RegisterSuccessor(BasicBlock& next);
```

Comment:
I guess "RegisterFoo" is a widely used naming pattern here and elsewhere, so never mind.